src/client.ts

```diff
@@ -4319,9 +4319,17 @@ export class MatrixClient extends TypedEventEmitter<EmittedEvents, ClientEventHa
             signPromise = this.http.requestOtherUrl<IThirdPartySigned>(Method.Post, url);
         }
 
-        const queryString: Record<string, string | string[]> = {};
+        const queryParams: QueryDict = {};
         if (opts.viaServers) {
-            queryString["server_name"] = opts.viaServers;
+            queryParams.server_name = opts.viaServers;
+            switch (this.canSupport.get(Feature.MigrateServerNameToVia)) {
+                case ServerSupport.Stable:
+                    queryParams.via = opts.viaServers;
+                    break;
+                case ServerSupport.Unstable:
+                    queryParams["org.matrix.msc4156.via"] = opts.viaServers;
+                    break;
+            }
```

Comment:
This feels quite verbose. I wanted to use `replaceParam` but I think it doesn't work in this case because we probably don't want to send the `via` parameter if there is no server support.